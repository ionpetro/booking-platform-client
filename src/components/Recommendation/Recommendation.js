import {gsap} from 'gsap';
import {BeatLoader} from '@saeris/vue-spinners';
import Button from '../Button/Button.vue';

export default {
  name: 'recommendation',
  components: {
    'beat-loader': BeatLoader,
    'v-button': Button
  },
  data() {
    return {
      loading: true,
      // hardcoded for demo purposes
      perfectMatchId: 3
    };
  },
  methods: {
    viewUnit() {
      this.$router.push({name: 'units', params: {id: this.perfectMatchId}});
    },
    animateRecommendationComponent() {
      gsap.set('.recommendation__features__feature', {
        x: -50,
        opacity: 0,
        transformOrigin: '50% 50%'
      });
      const tl = gsap.timeline({defaults: {duration: 1}});
      tl.to(
        '.recommendation__features__feature',
        {
          x: 0,
          opacity: 1,
          stagger: 0.8
        },
        '+=1'
      ).add(() => {
        this.loading = false;
      });
    }
  },
  mounted() {
    this.animateRecommendationComponent();
  }
};
