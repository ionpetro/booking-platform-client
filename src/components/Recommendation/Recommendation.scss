.recommendation {
  display: flex;
  justify-content: space-around;
  max-width: 1837px;
  margin: auto;

  &__features {
    position: relative;
    padding: 400px;
    // show a specific area of the infinite elevating image
    clip-path: polygon(0 0, 100% 0%, 100% 100%, 0% 100%);

    &__trancate--top {
      z-index: 1;
      position: absolute;
      top: 0;
      left: 0;
      width: 100%;
      height: 60px;
      background-image: linear-gradient(white, rgba(255, 255, 255, 0.2));
    }

    &__trancate--bot {
      z-index: 1;
      position: absolute;
      bottom: -2px;
      left: 0;
      width: 100%;
      height: 60px;
      background-image: linear-gradient(360deg, white, rgba(255, 255, 255, 0.2));
    }

    // add vertical infinite elevating image - first image
    &:before {
      position: absolute;
      top: 0;
      left: 0;
      content: '';
      display: block;
      animation: 40s bg linear infinite;
      background-image: url('~@/assets/images/Data.png');
      background-repeat: no-repeat;
      height: 828px;
      width: 640px;
    }

    // add vertical infinite elevating image - second image
    &:after {
      position: absolute;
      top: 0;
      left: 0;
      content: '';
      display: block;
      background-image: url('~@/assets/images/Data.png');
      background-repeat: no-repeat;
      animation: 40s bg2 linear infinite;
      height: 828px;
      width: 640px;
    }

    &__feature {
      z-index: 1;
      position: absolute;
      background: rgba(255, 255, 255, 0.75);
      box-shadow: 0px 4px 20px rgba(0, 0, 0, 0.15);
      border-radius: 20px;
      backdrop-filter: blur(2px);
      width: 347px;
      height: 82px;
      display: flex;
      justify-content: center;
      align-items: center;

      @media only screen and (max-width: 1680px) {
        width: 200px;
        height: 50px;
      }
      @media #{$tablet} {
        width: 50%;
        height: 60px;
      }

      @media #{$mobile} {
        width: 80%;
      }

      span {
        font-size: 26px;
        @media only screen and (max-width: 1680px) {
          font-size: 18px;
        }
      }
    }

    @media only screen and (max-width: 1980px) {
      padding: 350px 400px;
      &:after,
      &::before {
        margin-left: 150px;
      }
    }
    @media only screen and (max-width: 1680px) {
      padding: 400px 250px;
      &:after,
      &::before {
        margin-left: 50px;
      }
    }

    @media #{$tablet} {
      width: 100%;
      box-sizing: border-box;
      padding: 0;
      height: calc(100vh - 71px);
      &:after,
      &::before {
        display: none;
      }
    }
  }

  &__unit {
    text-align: center;

    &__loader {
      width: 900px;
      @media only screen and (max-width: 1980px) {
        width: 650px;
      }
      @media only screen and (max-width: 1480px) {
        width: 550px;
      }
      @media #{$tablet} {
        display: none;
      }
    }

    &__desc {
      &__img {
        @media only screen and (max-width: 1980px) {
          width: 650px;
        }
        @media only screen and (max-width: 1480px) {
          width: 550px;
        }
        @media #{$tablet} {
          width: 612px;
        }
        @media #{$mobile} {
          width: 272px;
        }
      }

      &__button {
        margin: auto;
        margin-top: 64px;
        @media #{$mobile} {
          margin-top: 24px;
        }
      }

      &__h1 {
        font-size: 48px;

        @media #{$mobile} {
          font-size: 24px;
        }
      }
      &__p {
        font-size: 18px;
        @media #{$mobile} {
          font-size: 14px;
        }
        margin: auto 24px;
      }

      @media #{$mobile} {
        .break {
          display: none;
        }
      }
    }
  }
}

@media #{$tablet} {
  .tablet__display {
    display: block;
  }
  .tablet__hide {
    display: none;
  }
}

.p1 {
  bottom: 82%;
  left: 20%;
  @include center-tablet-absolute-elements;
}

.p2 {
  bottom: 60%;
  left: 10%;
  @include center-tablet-absolute-elements;
}

.p3 {
  bottom: 38%;
  left: 50%;
  @include center-tablet-absolute-elements;
}

.p4 {
  bottom: 16%;
  left: 20%;
  @include center-tablet-absolute-elements;
}

@keyframes bg {
  0% {
    -webkit-transform: translateY(0);
    transform: translateY(0);
  }
  100% {
    -webkit-transform: translateY(-100%);
    transform: translateY(-100%);
  }
}
@keyframes bg2 {
  0% {
    -webkit-transform: translateY(100%);
    transform: translateY(100%);
  }
  100% {
    -webkit-transform: translateY(0%);
    transform: translateY(0%);
  }
}
