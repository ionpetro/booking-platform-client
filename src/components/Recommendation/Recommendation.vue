<template>
  <div class="recommendation">
    <div :class="['recommendation__features', loading ? 'tablet__display' : 'tablet__hide']">
      <div class="recommendation__features__trancate--top"></div>
      <div class="recommendation__features__trancate--bot"></div>
      <div class="recommendation__features__feature p1">
        <span>🥂 Definitely luxurious</span>
      </div>
      <div class="recommendation__features__feature p2"><span>📡 Fast connections </span></div>
      <div class="recommendation__features__feature p3"><span>🪐 Stunning view</span></div>
      <div class="recommendation__features__feature p4">
        <span>👽 Extra alien security</span>
      </div>
    </div>
    <div class="recommendation__unit">
      <div class="recommendation__unit__loader">
        <beat-loader v-if="loading" color="#0F1A3A" :size="20" sizeUnit="px"></beat-loader>
      </div>
      <transition name="fade">
        <div class="recommendation__unit__desc" v-if="!loading">
          <img
            class="recommendation__unit__desc__img"
            src="@/assets/images/RecommendationUnit.png"
            alt="Recommendation Unit"
          />
          <h1 class="recommendation__unit__desc__h1">Your perfect match!</h1>
          <p class="recommendation__unit__desc__p">
            After analyzing your thoughts and emotions, we found your ideal apartment!
            <br class="break" />
            Bright, luxurius with wide areas to invite your colleagues and friends.
          </p>
          <v-button
            @click.native="viewUnit"
            class="recommendation__unit__desc__button"
            :variant="'secondary'"
            >View this unit</v-button
          >
        </div>
      </transition>
    </div>
  </div>
</template>

<script src="./Recommendation.js"></script>
<style lang="scss" src="./Recommendation.scss" scoped></style>
