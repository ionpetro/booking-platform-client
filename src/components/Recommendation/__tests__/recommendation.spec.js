import {mount} from '@vue/test-utils';
import flushPromises from 'flush-promises';
import Recommendation from '../Recommendation.vue';

describe('Recommendation', () => {
  let wrapper;
  const $router = {push: jest.fn()};

  beforeEach(() => {
    wrapper = mount(Recommendation, {
      mocks: {$router}
    });
    // mount component and login a user
    wrapper.setData({loading: false});
  });

  describe('when loaded', () => {
    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
    it('redirects to perfect unit', async () => {
      await wrapper.find('.recommendation__unit__desc__button').trigger('click');

      await flushPromises();
      expect($router.push).toHaveBeenCalledWith({name: 'units', params: {id: 3}});
    });
  });
});
