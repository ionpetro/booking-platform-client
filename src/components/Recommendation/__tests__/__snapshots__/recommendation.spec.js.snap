// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recommendation when loaded renders correctly 1`] = `
<div
  class="recommendation"
>
  <div
    class="recommendation__features tablet__hide"
  >
    <div
      class="recommendation__features__trancate--top"
    />
     
    <div
      class="recommendation__features__trancate--bot"
    />
     
    <div
      class="recommendation__features__feature p1"
    >
      <span>
        🥂 Definitely luxurious
      </span>
    </div>
     
    <div
      class="recommendation__features__feature p2"
    >
      <span>
        📡 Fast connections 
      </span>
    </div>
     
    <div
      class="recommendation__features__feature p3"
    >
      <span>
        🪐 Stunning view
      </span>
    </div>
     
    <div
      class="recommendation__features__feature p4"
    >
      <span>
        👽 Extra alien security
      </span>
    </div>
  </div>
   
  <div
    class="recommendation__unit"
  >
    <div
      class="recommendation__unit__loader"
    >
      <!---->
    </div>
     
    <transition-stub
      name="fade"
    >
      <div
        class="recommendation__unit__desc"
      >
        <img
          alt="Recommendation Unit"
          class="recommendation__unit__desc__img"
          src="@/assets/images/RecommendationUnit.png"
        />
         
        <h1
          class="recommendation__unit__desc__h1"
        >
          Your perfect match!
        </h1>
         
        <p
          class="recommendation__unit__desc__p"
        >
          
          After analyzing your thoughts and emotions, we found your ideal apartment!
          
          <br
            class="break"
          />
          
          Bright, luxurius with wide areas to invite your colleagues and friends.
        
        </p>
         
        <button
          class="recommendation__unit__desc__button button secondary"
        >
          View this unit
        </button>
      </div>
    </transition-stub>
  </div>
</div>
`;
